Bundle refactor - `Option` bundles and `Box<dyn Bundle>`

SkiFire13 opened the pull request:
**Note: this is in the process of being split up, consider reviewing the individual PRs unless you're looking for the whole context**

# Objective

- Fixes #2157 by implementing `Bundle` for `Option<T: Bundle>`
- Fixes #3227 by implement `Bundle` for `Box<dyn Bundle>` and `Vec<Box<dyn Bundle>>`
- Supersedes #14879
- Don't regress the performance for current static bundles
- have `Option` bundles be faster than calling `insert` a second time
- Bonus: fix a bug where multiple `#[bundle(ignore)]` fields were handled in the wrong way

## Solution

- Introduce a `StaticBundle` (name bikesheddable) for usecases where the set of components of a bundle must be knowable without an instance of the bundle and switch functions that needed this aspect of `Bundle`;
- Add `&self` parameters to `Bundle`'s methods to allow them to depend on `self`'s value;
- (Optimization): add a way to distinguish different types of `Bundle`s:
  - _static_ bundles have all components statically known, like current bundles, and can be cached by their type id;
  - _bounded_ bundles may have some optional components, but overall the set of components they can contain is bounded, allowing to distinguish different subsets by a _cache key_ (in addition to the bundle type id);
  - _dynamic_ bundles have no way to be cached other than looking at their `ComponentId`s
  - overall this is implemented through the new `is_static`, `is_bounded` and `cache_key` methods on `Bundle`.
- Implement `Bundle` for `Option<T: Bundle>` as a _bounded_ bundle (i.e. with `is_static = false`, `is_bounded = true` and a cache key depending on whether it is `Some` or `None`);
- Make `Bundle` dyn-compatible and implement `Bundle` for `Box<dyn Bundle>`
  - this was done by marking non-dyn compatible methods as `where Self: Sized` and by introducing a couple of new dyn-compatible automatically-implemented supertraits, `BundleDyn` and `BundleEffectDyn`.
- Add a `#[bundle(dynamic)]` attribute for the `Bundle` derive macro to opt-out of deriving `StaticBundle` and `BundleFromComponents`
  - `BundleFromComponents` could theoretically be implemented for bounded but non-static bundles, but it seems tricky and I decided to leave it out from this PR

## Testing

- each new implementation of `Bundle` comes with its own tests in `crates/bevy_ecs/src/bundle.rs`
- benchmarks have been added to validate the performance advantages of the cache key mechanism, see [this](https://github.com/bevyengine/bevy/pull/19491#issuecomment-2952413782) and [this](https://github.com/bevyengine/bevy/pull/19491#issuecomment-2953985096) comments

---

## Showcase

```rs
fn enemy_bundle(name: String, attack_power: u32, is_boss: bool) -> impl Bundle {
    (
        EnemyMarker,
        Name::new(name),
        AttackPower(attack_power),
        // You can now conditionally insert components or other bundles!
        if is_boss { Some(BossMarker) } else { None }
    )
}

// You can now convert reflected components and bundles into an actual `Bundle` you can insert
fn bundle_from_reflected_data(components: Vec<Box<dyn Reflect>>, registry: &TypeRegistry) -> impl Bundle {
    components
        .into_iter()
        .map(|data| {
            let Some(reflect_bundle) = registry.get_type_data::<ReflectBundle>(data.type_id()) else { todo!() };
            let Ok(bundle_box) = reflect_bundle.get_boxed(data) else { todo!() };
            bundle_box
        })
        .collect::<Vec<Box<dyn Bundle>>>()
}
```

SkiFire13 committed:
Introduce StaticBundle

SkiFire13 committed:
Switch Entity{Ref,Mut}Except to StaticBundle

SkiFire13 committed:
Switch register_bundle to StaticBundle

SkiFire13 committed:
Switch EntityClonerBuilder to StaticBundle

SkiFire13 committed:
Switch Observer and Trigger to StaticBundle

SkiFire13 committed:
Add IS_STATIC and IS_BOUNDED flags to Bundle

SkiFire13 committed:
Add cache_key method to Bundle

SkiFire13 committed:
Add component_ids taking &self to Bundle

SkiFire13 committed:
Add register_required_components taking &self to Bundle

SkiFire13 committed:
Rename Bundles::bundle_ids to Bundles::static_bundle_ids

SkiFire13 committed:
Rename Bundles::register_info to register_static_info and add more dynamic register_info

SkiFire13 committed:
Convert BundleInserter to using the new register_info

SkiFire13 committed:
Switch component removal to StaticBundle

SkiFire13 committed:
Require StaticBundle in batch spawning

SkiFire13 committed:
Switch EntityWorldMut::retain to StaticBundle

SkiFire13 committed:
Rename World::register_bundle to World::register_static_bundle

SkiFire13 committed:
Rename register_contributed_bundle_info to register_static_contributed_bundle_info

SkiFire13 committed:
Add missing safety comments

SkiFire13 committed:
Require StaticBundle in ReflectBundle

SkiFire13 committed:
No longer require Bundle: StaticBundle

SkiFire13 committed:
Require StaticBundle in ExtractComponent

SkiFire13 committed:
Improve error message for an invalid dynamic Bundle

SkiFire13 committed:
WIP Implement Bundle for Option<T: Bundle>

SkiFire13 committed:
Simplify Bundle macro implementation

SkiFire13 committed:
Fix bug when #[bundle(ignore)] is followed by another one

SkiFire13 committed:
Add support for #[bundle(dynamic)] attribute

SkiFire13 committed:
Add simple test for Option bundles

SkiFire13 committed:
Switch IS_STATIC and IS_BOUNDED to associated functions

SkiFire13 committed:
Rename DynamicBundle to ComponentsFromBundle

SkiFire13 committed:
(WIP) Implement Bundle for Box<dyn Bundle>

SkiFire13 committed:
Fix formatting & doc issues

SkiFire13 committed:
Fix other CI issues

SkiFire13 committed:
Improve docs and safety comments

SkiFire13 committed:
Simplify Bundle derive implementation a bit more

SkiFire13 committed:
Fix typo

alice-i-cecile commented:
> I'm (personally) most interested in the Box<dyn Bundle> stuff (as allowing the insertion of a boxed bundle makes using them ergonomically viable for normal spawn scenarios).

This is my feeling too, but I won't object to the Option stuff either, as long as the perf isn't dangerously bad.

OleStrohm commented:
People are currently using optional Components, but in very unergonomic ways. There's a variety of ways to use the Children::spawn(SpawnIter(...)) (with Option::into_iter) and the SpawnableList (with a vec of zero or more elements) methods to get optional components for children (or an optional number of children), but this isn't currently possible for the top-level Commands::spawn, which means that people need to duplciate the spawning code for the case with and without the optional components (or spawn most of the components, and then Commands::insert the rest, which causes another archetype move).

In the last week I've seen at least three requests for optional bundles, and people expecting Option<B: Bundle> to just work, so there's a lot of demand for this.

So I think Option implementing Bundle is very valuable, and definitely going to be faster than the workarounds that currently exist.

tigregalis commented:
This is very cool and very much needed. I haven't looked at the implementation or given it a test-run just yet.

Is there a path from here to leptos-like control flow APIs and/or derive-on-enum as described in [this comment](https://github.com/bevyengine/bevy/issues/3227#issuecomment-2924924841)? I imagine enum bundles could take advantage of the same optimisations as `Option<Bundle>`.

With this implementation is it now possible to do:
```rs
fn returns_dynamic(a: bool, b: bool) -> impl Bundle {
    // type erase all bundles
    match (a, b) {
        (true, true) => Box::new((A, B)),
        (true, false) => Box::new(A),
        (false, true) => Box::new(B),
        (false, false) => Box::new(()),
    }
}
```
(I understand in this exact instance you would likely prefer `(Option<A>, Option<B>)`)

Can a fully dynamic bundle (i.e. a bundle "builder" with `insert()` methods) be built on top of this? Although I guess you could model this with  `#[derive(Builder)]` and `Option<Bundle>` in most instances.

SkiFire13 commented:
> I imagine enum bundles could take advantage of the same optimisations as `Option<Bundle>`.

Yes, the only reason this isn't done yet is deciding how to expose the interface for this (e.g. maybe expanding the derive macro to support enums?)

> Can a fully dynamic bundle (i.e. a bundle "builder" with insert() methods) be built on top of this? 

I haven't actually implemented this yet, but I think it should be possible to implement `Bundle` for `Vec<Box<dyn Bundle>>`, in which case your builder's `insert` would be just a `push` to this `Vec`.

I would still recommend using `Option` bundles over fully dynamic bundles for performance reasons though.

SkiFire13 committed:
Implement Bundle for Vec<Box<dyn Bundle>>

SkiFire13 committed:
Add some tests

tigregalis commented:
> Yes, the only reason this isn't done yet is deciding how to expose the interface for this (e.g. maybe expanding the derive macro to support enums?)
The two options are:


1. derive-on-enum

```rs
fn returns_custom(a: bool, b: bool) -> impl Bundle { // actual return type: Custom;
    #[derive(Bundle)]
    enum Custom {
        Ab((A, B)),
        A(A),
        B(B),
        None(()),
    }

    match (a, b) {
        (true, true) => Custom::Ab((A, B)),
        (true, false) => Custom::A(A),
        (false, true) => Custom::B(B),
        (false, false) => Custom::None(()),
    }
}
```

2. generic `EitherOfN` types

```rs
fn returns_either(a: bool, b: bool) -> impl Bundle { // actual return type: EitherOf4<(A, B), A, B, ()>;
    // provided the construction is exhaustive, Rust infers the return type
    match (a, b) {
        (true, true) => EitherOf4::A((A, B)),
        (true, false) => EitherOf4::B(A),
        (false, true) => EitherOf4::C(B),
        (false, false) => EitherOf4::D(()),
    }
}
```

I personally see `EitherOfN` as the more ergonomic option as it doesn't *require* you to write out a transient type. Not precisely sure how well you could extract the different component-ID-sets at compile time however. At runtime though it should be possible to inspect the type and delegate to the inner type's component-ID-set. Not sure how well that fits in with your `Option<B: Bundle>` design.

Benefit with the derive is you probably could get all the information at compile time more easily.

They don't have to be mutually exclusive of course but it does require justifying the increased API surface.

> I think it should be possible to implement `Bundle` for `Vec<Box<dyn Bundle>>`, in which case your builder's `insert` would be just a `push` to this `Vec`.

This brings to mind `[Box<dyn Bundle>; N]` and `Box<[Box<dyn Bundle>]>`, but I guess they are strictly worse than `(B1, B2, B3, ..)` and `Vec<Box<dyn Bundle>>` (except size) respectively, and of course they are convertible to/from `Vec`.

SkiFire13 commented:
> I personally see EitherOfN as the more ergonomic option as it doesn't require you to write out a transient type. Not precisely sure how well you could extract the different component-ID-sets at compile time however. At runtime though it should be possible to inspect the type and delegate to the inner type's component-ID-set. Not sure how well that fits in with your Option<B: Bundle> design.

It should work fine, my design does not need to know the components at compile time. However for the optimized path it needs to know that they are *bounded* at compile time.

> Could you compute the cache key or is that compile-time information?

For anything involving `Box<dyn Bundle>` it's not possible, as that is not bounded.

SkiFire13 committed:
Add benchmarks for bundles

SkiFire13 commented:
I've added benchmarks for bundle insertions, here are the results on my Macbook M4 Pro 24GB:

```
ecs::bundles::spawn_many::spawn_many/box_dyn_bundle_many                       552.0±14.01µs
ecs::bundles::spawn_many::spawn_many/dynamic_bundle_many                       450.3±28.67µs
ecs::bundles::spawn_many::spawn_many/many_box_dyn_bundle                      1457.9±17.64µs
ecs::bundles::spawn_many::spawn_many/many_dynamic_bundle                       442.7±33.86µs
ecs::bundles::spawn_many::spawn_many/option_many_none_and_static                132.1±1.44µs
ecs::bundles::spawn_many::spawn_many/option_many_some                           135.7±3.41µs
ecs::bundles::spawn_many::spawn_many/option_none_many                            40.1±0.37µs
ecs::bundles::spawn_many::spawn_many/option_some_many                           139.2±2.83µs
ecs::bundles::spawn_many::spawn_many/static                                     118.8±2.51µs
ecs::bundles::spawn_many_zst::spawn_many_zst/box_dyn_bundle_many               516.7±34.06µs
ecs::bundles::spawn_many_zst::spawn_many_zst/dynamic_bundle_many               412.9±31.14µs
ecs::bundles::spawn_many_zst::spawn_many_zst/many_box_dyn_bundle               610.7±52.70µs
ecs::bundles::spawn_many_zst::spawn_many_zst/many_dynamic_bundle               400.3±28.39µs
ecs::bundles::spawn_many_zst::spawn_many_zst/option_many_none_and_static         98.4±5.90µs
ecs::bundles::spawn_many_zst::spawn_many_zst/option_many_some                   102.0±3.49µs
ecs::bundles::spawn_many_zst::spawn_many_zst/option_none_many                    39.4±0.32µs
ecs::bundles::spawn_many_zst::spawn_many_zst/option_some_many                   109.5±2.82µs
ecs::bundles::spawn_many_zst::spawn_many_zst/static                              89.4±6.22µs
ecs::bundles::spawn_one_zst::spawn_one_zst/box_dyn_bundle                       565.4±8.16µs
ecs::bundles::spawn_one_zst::spawn_one_zst/dynamic_bundle                       381.7±7.08µs
ecs::bundles::spawn_one_zst::spawn_one_zst/option_none                          417.4±9.75µs
ecs::bundles::spawn_one_zst::spawn_one_zst/option_none_and_static              399.0±14.06µs
ecs::bundles::spawn_one_zst::spawn_one_zst/option_some                         382.1±10.34µs
ecs::bundles::spawn_one_zst::spawn_one_zst/static                               156.1±7.97µs
```

I benched 3 scenarios:
- spawning a single ZST component: this is supposed to show the base overhead of `Option` and dynamic bundles.
  - `Option` bundles and dynamic bundles seem to be 2x slower than static bundles
  - `Box<dyn Bundle>` is almost 4x slower, likely due to needing dynamic dispatch (note: there are no allocations here because I benched with ZSTs)
  - interestingly dynamic bundles are on par if not slightly faster than `Option` bundles in this case, I'm not sure why this's the case;
- spawning many ZST components: this is supposed to show how the overhead scales with more sub-bundles, though this is also a bit amortized by the fact we need to actually insert more components.
  - the overhead of `Option` bundles compared to static bundles is now at around 10-20%, likely because inserting the components is taking more time overall;
  - dynamic bundles remain 4-6x slower depending on how much dynamic dispatch there's overall;
- spawning many non-ZST components: this is supposed to be a more realistic example, where `Box<dyn Bundle>`s have to allocate and the cost of actually inserting the components is higher:
  - `Option` bundles's overhead are still at 10-20%
  - dynamic bundles remain 4-5x slower with the exception of spawning many `Box<dyn Bundle>`, where the overhead of allocating `Box`es makes it much slower.

Overall I think 10-20% overhead for `Option` bundles is pretty good, although I feel like it could be better. Dynamic bundles being 4-5x slower is not optimal however, especially considering it's enough to have one dynamic sub-bundle to make your bundle also dynamic and suffer from the overall slowdown.

SkiFire13 committed:
Add BoundedBundleKey to simplify implementations of cache_key

SkiFire13 committed:
Add migration guide

SkiFire13 committed:
Add release notes

SkiFire13 committed:
Implement ReflectBundle::get_boxes

alice-i-cecile commented:
I'd love to see benchmarks for chained / conditional .insert calls: to me that's the real alternative here in terms of performance. We can and should split those up from this PR: those are useful benchmarks for our later command batching work anyways. I still think we should proceed with this regardless of the slow perf, but it's good to be able tot test / communicate what the alternatives are.

SkiFire13 committed:
Add more benchmarks

SkiFire13 commented:
I added a couple more benchmarks to compare with separate `insert`s, in particular:

- `insert_many/only_last`, where I `spawn` a bunch of components and then `insert` the last one
- `spawn_many/option_one_none`, where I spawn a bunch of components together with a single `None`
- `spawn_many/option_one_some`, where I spawn a bunch of components together with a single `Some`

These are supposed to simulate the insertion of an optional component, which is currently done with e.g. an `insert_if` command which performs an additional `insert`. The results shows that inserting an `Option` bundle instead is ~2x faster. Note though that the `insert_many/only_last` emulates the case where the optional component is present. If it isn't then there would be no `insert` and a result more similar to `spawn_many/static` is expected (though now that I'm writing this I realized that `spawn_many/static` is spawning one more component, so it's not a proper comparison).

```
ecs::bundles::insert_many::insert_many/only_last                                238.1±2.04µs
ecs::bundles::spawn_many::spawn_many/option_one_none                            118.6±1.79µs
ecs::bundles::spawn_many::spawn_many/option_one_some                            124.9±2.17µs
```

For completeness I also added a benchmark where an `insert` for each component is called, and as we can expect this is pretty slow:

```
ecs::bundles::insert_many::insert_many/all                                    1273.3±12.22µs
```

@alice-i-cecile

alice-i-cecile commented:
Lovely; that's helpful justification, and those benches will be useful in the future.

SkiFire13 commented:
I think the `S-Needs-Help`, `S-Waiting-on-Author` and `S-Needs-Benchmarking` labels could be removed now as the PR should now be ready for review and include benchmarks. I'm not sure how these are generally managed though.

SkiFire13 commented:
Upon seeing #19560 I wonder if `StaticBundle` should be a separate derive or not. Differently from `BundleFromComponents`, which is just used in `take`, `StaticBundle` is used in a lot of places (almost every API that uses `Bundle` but not for `insert` or `spawn`) so a double derive might be a lot more common. For reference, currently I have a `#[bundle(dynamic)]` annotation to opt-out of deriving `StaticBundle` and `BundleFromComponents`.

alice-i-cecile commented:
I think that making the common case pleasant is the way to go here. Most bundles will be static, so I like your current design.

SkiFire13 committed:
Merge remote-tracking branch 'upstream/main' into new-bundle

SkiFire13 committed:
Merge remote-tracking branch 'upstream/main' into new-bundle

SkiFire13 committed:
Merge remote-tracking branch 'upstream/main' into new-bundle

SkiFire13 committed:
Merge remote-tracking branch 'upstream/main' into new-bundle

SkiFire13 committed:
Fix imports and re-exports

SkiFire13 committed:
Merge remote-tracking branch 'upstream/main' into new-bundle

SkiFire13 committed:
Merge remote-tracking branch 'upstream/main' into new-bundle
